src/Compiler/Checking/PostInferenceChecks.fs

```diff
@@ -432,11 +432,11 @@ and CheckTypeConstraintDeep cenv f g env x =
      | TyparConstraint.IsReferenceType _
      | TyparConstraint.RequiresDefaultConstructor _ -> ()
 
-and CheckTraitInfoDeep cenv (_, _, _, visitTraitSolutionOpt, _ as f) g env (TTrait(tys, _, _, argTys, retTy, soln))  =
-    CheckTypesDeep cenv f g env tys
-    CheckTypesDeep cenv f g env argTys
-    Option.iter (CheckTypeDeep cenv f g env true ) retTy
-    match visitTraitSolutionOpt, soln.Value with
+and CheckTraitInfoDeep cenv (_, _, _, visitTraitSolutionOpt, _ as f) g env traitInfo =
+    CheckTypesDeep cenv f g env traitInfo.SupportTypes
+    CheckTypesDeep cenv f g env traitInfo.CompiledObjectAndArgumentTypes
+    Option.iter (CheckTypeDeep cenv f g env true ) traitInfo.CompiledReturnType
+    match visitTraitSolutionOpt, traitInfo.Solution with
```

Comment:
Why specifically rewrite as property call as opposed to deconstructing?

Reply:
Just readability. I can change it back to deconstruction. Do you prefer that?

Reply:
I don't mind either, was just curious

Reply:
@Martin521 I think the potential issue is, unless the JIT optimizes it, accessing more than a single member means doing the check on the tag for each property access.

It is the same when extracting a single value, but when there are repeated accesses, it does need to check the tag again.

Reply:
Jit will likely optimize it (also will likely inline thongs), I wouldn't worry about significant perf issues.

Reply:
@smoothdeveloper thanks for the comment, it made me learn something again. Because I first just wanted to agree and revert to deconstruction. But then my curiosity was too strong and I [checked the IL](https://sharplab.io/#v2:DYLgZgzgNALiCWwoBMQGoA+MCeAHApgAQAqAwgJKEC8JxhA9mIQIYiHwB2MhAVIQEZswwes258AxmwgwATpwDmhAO7wYACwCwAKEJ7CAW3wH++WYRgA6AILVCwfNwAUxYk+Z3mASi93unFh19Q2NTcysAITsHZ1cnfjt+Hz92DgEg/SMTMwtLUmjHQhc3CTsJZJp/NIkdHRjCXABGIrgSCl8qDL1ceS4wNIAiAFJkQiGmIYgB3NtI3NI6wtwAJiK45igBTfKOrobemH7CYdHxsamWAUIa7SA). At least in that simplified example, both versions have more or less the same IL, the version using multiple property access is actually even a little smaller and faster since it saves a few register moves.

Reply:
Interesting, it feels like there could be an optimization step that elides locals that are only storing field access into locals and there are no writes.

Thanks for looking into it!

Reply:
I am not so fluent in assembly, but it looks like JIT actually does this optimization.
Definitely, the two versions are exactly the same after JIT.